### ISSUE TITLE
question on nl-lassysmall

### ISSUE THREAD

**Issue by jwijffels** [ufal/udpipe#57]
I'm trying to reproduce the models released with ud-2.0 using the parameters defined at https://github.com/ufal/udpipe/tree/master/training/models-ud-2.0 
In particular, I'm have a question on nl-lassysmall. The model does not output xpos tags (checked this on the R side but also in the webservice at http://lindat.mff.cuni.cz/services/udpipe/).
`
![knipsel](https://user-images.githubusercontent.com/1710810/34611828-65eba812-f227-11e7-9d9e-0fb9b7251fd2.PNG)
`
While if I train the model from R with the following parameters using the latest conllu files (version 2.1), I do get the xpos tags. 
I checked that the xpos tags were there in version 2.0 (March 1) and they were not available, so that explains probably it. 
However, to understand the number reported in http://dx.doi.org/10.18653/v1/K17-3009 for the xpos for dutch-lassysmall (99.9%). Can you explain where does it come from.
In particular, I'm interested basically to understand if there any further things that were done except the code at https://github.com/ufal/udpipe/tree/master/training (resplitting data, computing embeddings and training the model with the provided parameters) to the models released at https://lindat.mff.cuni.cz/repository/xmlui/handle/11234/1-2364.

```
library(udpipe)
m <- udpipe_train(file = "dev/nl-lassysmall-token-tag-parse.udpipe", 
                  files_conllu_training = "dev/nl_lassysmall-ud-train.conllu", 
                  files_conllu_holdout = "dev/nl_lassysmall-ud-test.conllu",
                  annotation_tokenizer = list(tokenize_url = 1, allow_spaces = 0,
                                              dimension = 64, epochs = 100, initialization_range = 0.1,
                                              batch_size = 100, learning_rate = 0.005, dropout = 0.1, early_stopping = 1),
                  annotation_tagger = list(models = 2, iterations = 20,
                                           ## Settings for the UPOS/XPOS/FEATS tagger
                                           templates_1 = "tagger", 
                                           guesser_suffix_rules_1 = 10, guesser_enrich_dictionary_1 = 6, guesser_prefixes_max_1 = 1,
                                           use_lemma_1 = 1, use_xpostag_1 = 1, use_feats_1 = 1, 
                                           provide_lemma_1 = 0, provide_xpostag_1 = 1, provide_feats_1 = 1, prune_features_1 = 0,
                                           ## Settings for the Lemmatizer
                                           templates_2 = "lemmatizer",
                                           guesser_suffix_rules_2 = 8, guesser_enrich_dictionary_2 = 5, guesser_prefixes_max_2 = 4,
                                           use_lemma_2 = 1, use_xpostag_2 = 0, use_feats_2 = 0, 
                                           provide_lemma_2 = 1, provide_xpostag_2 = 0, provide_feats_2 = 0, prune_features_2 = 1), 
                  annotation_parser = list(iterations = 30, 
                                           embedding_upostag = 20, embedding_feats = 20, embedding_xpostag = 0, 
                                           embedding_form = 50, embedding_form_file = "dev/ud-2.0-embeddings/nl_lassysmall.skip.forms.50.vectors", 
                                           embedding_lemma = 0, embedding_deprel = 20, 
                                           learning_rate = 0.01, learning_rate_final = 0.001, l2 = 0.3, hidden_layer = 200, 
                                           batch_size = 10, transition_system = "swap", transition_oracle = "static_lazy", 
                                           structured_interval = 8))
m <- udpipe_load_model("dev/nl-lassysmall-token-tag-parse.udpipe")
as.data.frame(udpipe_annotate(m, "Hoe komt dit, er zijn geen xpostags hierbij."))

   doc_id paragraph_id sentence_id                                     sentence token_id    token   lemma  upos                                 xpos
1    doc1            1           1 Hoe komt dit, er zijn geen xpostags hierbij.        1      Hoe     hoe   ADV                                   BW
2    doc1            1           1 Hoe komt dit, er zijn geen xpostags hierbij.        2     komt   komen  VERB                      WW|pv|tgw|met-t
3    doc1            1           1 Hoe komt dit, er zijn geen xpostags hierbij.        3      dit     dit  PRON         VNW|aanw|pron|stan|vol|3o|ev
4    doc1            1           1 Hoe komt dit, er zijn geen xpostags hierbij.        4        ,       , PUNCT                                  LET
5    doc1            1           1 Hoe komt dit, er zijn geen xpostags hierbij.        5       er      er   ADV   VNW|aanw|adv-pron|stan|red|3|getal
6    doc1            1           1 Hoe komt dit, er zijn geen xpostags hierbij.        6     zijn    zijn   AUX                         WW|pv|tgw|mv
7    doc1            1           1 Hoe komt dit, er zijn geen xpostags hierbij.        7     geen    geen   DET VNW|onbep|det|stan|prenom|zonder|agr
8    doc1            1           1 Hoe komt dit, er zijn geen xpostags hierbij.        8 xpostags xpostag  NOUN                     N|soort|mv|basis
9    doc1            1           1 Hoe komt dit, er zijn geen xpostags hierbij.        9  hierbij hierbij   ADV                                   BW
10   doc1            1           1 Hoe komt dit, er zijn geen xpostags hierbij.       10        .       . PUNCT                                  LET
```

**Comment by jwijffels** [comment]
Right. Found it. Excuse me for filling the issue. I had to look here: http://ufal.mff.cuni.cz/udpipe/users-manual#universal_dependencies_20_models for looking at model statistics.

**Comment by foxik** [comment]
The model released at http://hdl.handle.net/11234/1-2364 are trained exactly as described in the udpipe-ud-2.0-170801-reproducible_training.zip file present on the link -- as you say, we resplit the data, compute the embeddings [actually already precomputed in the arxive] and train the models with the parameters from the params_* files.

As for the XPOS tags -- it is exactly as you say, in 2.0 there were no XPOS tags. The evaluation of the models present at http://ufal.mff.cuni.cz/udpipe/users-manual#conll17_shared_task_baseline_ud_20_models reflects this and marks the XPOS tags of Dutch-LassySmall as not present. However, the CoNLL 17 evaluation explicitly lists all annotations, giving 100% to the "empty" ones -- but the results are reported word-aligned. Therefore, because the word alignment of the automatic Dutch-LassySmall tokenizer is 99.9%, the XPOS is also 99.9%.

**Comment by jwijffels** [comment]
Thanks for the input. Indeed the numbers at http://ufal.mff.cuni.cz/udpipe/users-manual#universal_dependencies_20_models are the right ones. 
Let me try to reproduce by taking the test data from ud 2.0 and seeing what the output of the annotate function is.
Shown below from the R side. The xpostag mentions xpostag: 99.96%. (that is what is printed using ufal::udpipe::evaluator::evaluate). So to understand this correctly, the evaluation of xpos does not make sense. And that is why you are using something like has_xpos in https://github.com/ufal/udpipe/blob/master/training/models-ud-2.0/results_table.sh?

```
library(utils)
library(udpipe)
download.file(url = "https://raw.githubusercontent.com/UniversalDependencies/UD_Dutch-LassySmall/3784b72a0553a849bb9a28763b71c30fde862276/nl_lassysmall-ud-test.conllu", 
              destfile = "test.conllu")
m <- udpipe_download_model("dutch-lassysmall")
m <- udpipe_load_model(m$file_model)
as.data.frame(udpipe_annotate(m, "Geen xpos tags"))
  doc_id paragraph_id sentence_id       sentence token_id token lemma upos xpos                  feats head_token_id dep_rel deps            misc
1   doc1            1           1 Geen xpos tags        1  Geen  geen  DET <NA>                   <NA>             3     det <NA>            <NA>
2   doc1            1           1 Geen xpos tags        2  xpos  xpos  ADJ <NA>             Degree=Pos             3    amod <NA>            <NA>
3   doc1            1           1 Geen xpos tags        3  tags   tag NOUN <NA> Gender=Com|Number=Sing             0    root <NA> SpacesAfter=\\n

evaluation <- udpipe_accuracy(m, "test.conllu")
cat(evaluation$accuracy, sep = "\n")

Number of SpaceAfter=No features in gold data: 1378
Tokenizer words - system: 10083, gold: 10080, precision: 99.94%, recall: 99.97%, f1: 99.96%
Tokenizer sentences - system: 789, gold: 800, precision: 80.99%, recall: 79.88%, f1: 80.43%
Tagging from plain text (CoNLL17 F1 score) - gold forms: 10080, upostag: 97.63%, xpostag: 99.96%, feats: 97.19%, alltags: 96.48%, lemmas: 98.07%
Parsing from plain text with computed tags (CoNLL17 F1 score) - gold forms: 10080, UAS: 84.37%, LAS: 81.98%
```

**Comment by foxik** [comment]
The numbers (xpos and alltags) do make sense -- but in case you start with a plain text (and not a pretokenized input), they are F1 scores described in detail at http://universaldependencies.org/conll17/evaluation.html. (That is why I write "CoNLL F1 score".) For missing features, you get 100% accuracy; but the final F1 results are a combination of word-alignment and accuracy. But even if they make sense, they are quite uninformative for missing features (i.e., the same as word-alignment F1), so that is why I am "removing" them in results_table.

The effect of missing XPOS on AllTags is none -- for AllTags, you first compute an accuracy for every word (getting "+1" if all tags match, i.e., UPOS, XPOS and FEATS, and "0" otherwise -- and if XPOS is missing, you are effectively comparing only UPOS and FEATS), and then combine it with the word-alignment F1.

Note that it is possible to get more "common" results by starting with already pretokenized text (i.e., the lines "Gold tok" at http://ufal.mff.cuni.cz/udpipe/users-manual#universal_dependencies_20_models) -- you can achieve it by evaluating only tagger and parser (i.e., passing Evaluator.NONE as tokenizer).

**Comment by jwijffels** [comment]
Ok. Thanks for the clarifications! Clear for me. 
I'm getting exactly these numbers as reported on http://ufal.mff.cuni.cz/udpipe/users-manual#universal_dependencies_20_models.

```
evaluation <- udpipe_accuracy(m, "test.conllu", tokenizer = "none")
cat(evaluation$accuracy, sep = "\n")
Tagging from gold tokenization - forms: 10080, upostag: 97.74%, xpostag: 100.00%, feats: 97.42%, alltags: 96.76%, lemmas: 98.15%
Parsing from gold tokenization with computed tags - forms: 10080, UAS: 87.50%, LAS: 85.02%
evaluation <- udpipe_accuracy(m, "test.conllu", tokenizer = "none", tagger = "none")
cat(evaluation$accuracy, sep = "\n")
Parsing from gold tokenization with gold tags - forms: 10080, UAS: 89.68%, LAS: 87.41%
```
